src/Service/Bulk/BulkCreateUsersService.php

```diff
@@ -0,0 +1,333 @@
+ *
+ *  Copyright (c) 2021 (original work) Open Assessment Technologies S.A.
+ */
+
+declare(strict_types=1);
+
+namespace OAT\SimpleRoster\Service\Bulk;
+
+use OAT\SimpleRoster\Csv\CsvWriter;
+use OAT\SimpleRoster\DataTransferObject\UserCreationResult;
+use OAT\SimpleRoster\DataTransferObject\UserCreationResultMessage;
+use OAT\SimpleRoster\Exception\LineItemNotFoundException;
+use OAT\SimpleRoster\Lti\Service\GenerateGroupIdsService;
+use OAT\SimpleRoster\Repository\AssignmentRepository;
+use OAT\SimpleRoster\Repository\Criteria\FindLineItemCriteria;
+use OAT\SimpleRoster\Repository\LineItemRepository;
+use OAT\SimpleRoster\Repository\LtiInstanceRepository;
+use Symfony\Component\Filesystem\Filesystem;
+
+class BulkCreateUsersService
+{
+    private array $lineItemIds = [];
+    private array $lineItemSlugs = [];
+
+    private int $userGroupBatchCount = 0;
+    private int $groupIndex = 0;
+
+    private string $generatedUsersFilePath;
+    private string $projectDir;
+
+    private const USER_CSV_HEAD = ['username', 'password', 'groupId'];
+    private const ASSIGNMENT_CSV_HEAD = ['username', 'lineItemSlug'];
+
+    private LineItemRepository $lineItemRepository;
+    private AssignmentRepository $assignmentRepository;
+    private Filesystem $filesystem;
+    private GenerateGroupIdsService $generateGroupIdsService;
+    private CsvWriter $csvWriter;
+    private LtiInstanceRepository $ltiInstanceRepository;
+    private UserCreationResultMessage $userCreationMessage;
+
+    private const DEFAULT_USERNAME_INCREMENT_VALUE = 0;
+
+    public function __construct(
+        LineItemRepository $lineItemRepository,
+        AssignmentRepository $assignmentRepository,
+        GenerateGroupIdsService $generateGroupIdsService,
+        CsvWriter $csvWriter,
+        LtiInstanceRepository $ltiInstanceRepository,
+        UserCreationResultMessage $userCreationMessage,
+        Filesystem $filesystem,
+        string $generatedUsersFilePath,
+        string $projectDir
+    ) {
+        $this->lineItemRepository = $lineItemRepository;
+        $this->assignmentRepository = $assignmentRepository;
+        $this->generateGroupIdsService = $generateGroupIdsService;
+        $this->userCreationMessage = $userCreationMessage;
+        $this->csvWriter = $csvWriter;
+        $this->generatedUsersFilePath = $generatedUsersFilePath;
+        $this->filesystem = $filesystem;
+        $this->ltiInstanceRepository = $ltiInstanceRepository;
+        $this->projectDir = $projectDir;
+    }
+
+    /**
+     * @throws LineItemNotFoundException
+     */
+    public function createUsers(
+        array $lineItemIds,
+        array $lineItemSlugs,
+        array $userPrefixes,
+        int $batchSize,
+        ?string $groupPrefix
+    ): UserCreationResult {
+        $notExistLineItemsArray = $userGroupIds = [];
+
+        if (!empty($lineItemIds) || !empty($lineItemSlugs)) {
+            $this->lineItemIds = $lineItemIds;
```

Comment:
I don't see why do we need to have `$this->lineItemIds` as a class parameter. we need it only here and in 
getFindLineItemCriteria. and we can pass it through parameters. right?

Reply:
and in this case getFindLineItemCriteria probably should be a separate factory. so it can be mocked and easily reused.